From ad74a85e0fa52d67331894bce9c532461302ed7d Mon Sep 17 00:00:00 2001
From: z00512904 <z00512904@huawei.com>
Date: Tue, 6 Sep 2022 14:31:28 +0000
Subject: [PATCH] [Huawei] using mm's pasid rather than a random value when
 using SDMA

Offering:ascend
CVE:
Reference:
Type:feature
reason: when using SDMA feature, we must using mm's pasid.

Signed-off-by: z00512904 <z00512904@huawei.com>
---
 drivers/char/sdma.c | 9 +++++----
 1 file changed, 5 insertions(+), 4 deletions(-)

diff --git a/drivers/char/sdma.c b/drivers/char/sdma.c
index 7380be992..7ac008046 100644
--- a/drivers/char/sdma.c
+++ b/drivers/char/sdma.c
@@ -833,7 +833,6 @@ struct file_open_data {
 static int __do_sdma_open(struct sdma_device *psdma_dev, struct file *file)
 {
 	void *ret;
-	int pasid;
 	struct file_open_data *data;
 
 	data = kmalloc(sizeof(*data), GFP_KERNEL);
@@ -848,7 +847,7 @@ static int __do_sdma_open(struct sdma_device *psdma_dev, struct file *file)
 	}
 
 	data->sva = ret;
-	data->pasid = pasid;
+	data->pasid = current->mm->pasid;
 	data->psdma_dev = psdma_dev;
 	file->private_data = data;
 
@@ -875,8 +874,10 @@ static int sdma_core_open(struct inode *inode, struct file *file)
 
 static int sdma_dev_release(struct inode *inode, struct file *file)
 {
-	/* We don't unbind current process since other device may use it */
-	kfree(file->private_data);
+	struct file_open_data *data = file->private_data;
+
+	iommu_sva_unbind_device(data->sva);
+	kfree(data);
 
 	return 0;
 }
-- 
2.17.1

